Resolve crash when submitting feedback without internet access

angrezichatterbox opened the pull request:
**Description (required)**

Fixes #6003 

What changes did you make and why?
When the internet connection was disabled, the feedback page on submission would crash. This was because there were no conditions that would gracefully handle the absence of an internet connection when the request was being sent. To address this issue, I implemented a check to verify the availability of the internet connection within the uploadFeedback function. If there was no active connection, We would exit the loop and display a Toast with an error description. 

**Tests performed (required)**

Tested prodRelease on Medium Phone with API level 35.

neeldoshii commented:
Hi @angrezichatterbox, thanks a lot for contributing!
Though checking whether you are on wifi or cellular does solve the problem, if u have explicitly closed cellular and wifi. But what if you are connected to cellular/wifi but the network is not proper/low or is not working for any reason. 
(as per system service) the code will work as you are connected to wifi/cellular and then it will crash.

I managed to break the app on your branch.
![image](https://github.com/user-attachments/assets/74a8f3e0-735e-4d61-857f-ea726888cbfa)

Steps to break.
1. Be connected to internet/wifi
2. But the internet is very slow or not able to perform communication with the server.
3. Submit Feedback.
4. Crash


What if we try to add an exception while performing an API call and if it throws error -> show user error as failed.

neeldoshii commented on app/src/main/java/fr/free/nrw/commons/navtab/MoreBottomSheetFragment.kt:
```diff
@@ -170,6 +171,11 @@ class MoreBottomSheetFragment : BottomSheetDialogFragment() {
         val feedbackContentCreator = FeedbackContentCreator(requireContext(), feedback)
 
+        if (!isNetworkAvailable(requireContext())) {
+            Toast.makeText(requireContext(), R.string.error_feedback, Toast.LENGTH_LONG).show()
```
Instead of toast, try having Snackbar with an action label (retry) to perform retry.

angrezichatterbox replied:
So I should make show a snack bar which would have the option to reshow the Feedback Dialog. Am I right ?

neeldoshii replied:
![image](https://github.com/user-attachments/assets/50469815-8a3a-4215-91b9-444ec15b997a)
Suggestion : Allow the user to perform retry if the feedback is not submitted. Wdyt?

nicolas-raoul replied:
Allowing the user to retry would be great indeed.
Maybe leave the form there until sent successfully?

angrezichatterbox commented:
Thanks for the review @neeldoshii. I would make the required changes and update the PR :)

angrezichatterbox commented:
I’ve made the required changes. Could you please review the PR again, @neeldoshii? Thanks a lot! :)

neeldoshii commented:
![image](https://github.com/user-attachments/assets/1d0c702d-acde-4f05-8589-46a8c2c6211e)
still crashes to me.

neeldoshii commented on app/src/main/java/fr/free/nrw/commons/navtab/MoreBottomSheetFragment.kt:
```diff
@@ -5,14 +5,24 @@ import android.annotation.SuppressLint
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkInfo
```
Unused Import

neeldoshii commented on app/src/main/java/fr/free/nrw/commons/navtab/MoreBottomSheetFragment.kt:
```diff
@@ -170,6 +181,28 @@ class MoreBottomSheetFragment : BottomSheetDialogFragment() {
     fun uploadFeedback(feedback: Feedback) {
         val feedbackContentCreator = FeedbackContentCreator(requireContext(), feedback)
 
+        if (!isInternetConnectionAvailable(requireContext())) {
```
The issue doesn't lies when you are communicating between client & server. The issue is we are not handling failing of network call when the problem arises due to low/no data connection, if I am not wrong.

neeldoshii commented on app/src/main/java/fr/free/nrw/commons/navtab/MoreBottomSheetFragment.kt:
```diff
@@ -182,16 +215,36 @@ class MoreBottomSheetFragment : BottomSheetDialogFragment() {
+            val networkCapabilities = connectivityManager.getNetworkCapabilities(activeNetwork)
+            networkCapabilities?.hasCapability(android.net.NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
+        } else {
+            val activeNetworkInfo: NetworkInfo? = connectivityManager.activeNetworkInfo
```
'NetworkInfo' is deprecated. Deprecated in Java
'NetworkInfo' is deprecated. Deprecated in Java
'getter for activeNetworkInfo: NetworkInfo?' is deprecated. Deprecated in Java
'getter for isConnected: Boolean' is deprecated. Deprecated in Java

angrezichatterbox replied:
I had to use because the current minimum API level was 21 and I had to use it for devices below API level 23

nicolas-raoul replied:
I don't think that checking the connection is the way to go, because:
1. It is complex.
2. Internet connection can be lost immediately after the apps checks it.

Instead I would recommend a try/catch approach:
1. Try to submit the feedback
2. If there was any kind of exception (might be internet connection error or server timeout or other issues), alert the user rather than crashing.

neeldoshii replied:
check this @angrezichatterbox 
```
try{
//for example, retrofit call
}
catch (e: Exception) {
        is UnknownHostException -> "Unknown host!"
        is ConnectException -> "No internet!"
        else -> "Unknown exception!"
    }
```

Ref : https://stackoverflow.com/a/61099415

angrezichatterbox commented:
> ![image](https://github.com/user-attachments/assets/1d0c702d-acde-4f05-8589-46a8c2c6211e)
> 
> still crashes to me, can you try to replicate this setting for your emulator and try having cellular on in your emulator.
> 
> 

I'll look into this and fix it.

angrezichatterbox commented:
I’ve made the necessary changes. Please let me know if I missed any. I’m happy to make further changes to ensure the PR is perfect.

neeldoshii commented:
Note : 
The first step now is to, try to crash the app once forcefully. Check the logs and try to find whats the root cause of it and how to fix it. As said by Nicolas (https://github.com/commons-app/apps-android-commons/pull/6018#discussion_r1879413241), checking internet connection before submitting might not be the correct way. Better approach would be to use try.. catch and in catch check the error exception and notify the user.

Do lmk if you need any help.
 
https://github.com/user-attachments/assets/74bab23b-ac74-44b6-b723-a8d4d2e6de03

Logs:-

```
java.io.IOException: {"errors":[{"code":"login-required","text":"You must be logged in.","module":"query+notifications"}],"docref":"See https://commons.wikimedia.org/w/api.php for API usage. Subscribe to the mediawiki-api-announce mailing list at &lt;https://lists.wikimedia.org/postorius/lists/mediawiki-api-announce.lists.wikimedia.org/&gt; for notice of API deprecations and breaking changes.","servedby":"mw-api-ext.codfw.main-f7bb96999-z8zc2"}
                                                                                        	at fr.free.nrw.commons.OkHttpConnectionFactory$UnsuccessfulResponseInterceptor.intercept(OkHttpConnectionFactory.java:99)
                                                                                        	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
                                                                                        	at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.kt:221)
                                                                                        	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
                                                                                        	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
                                                                                        	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
                                                                                        	at retrofit2.OkHttpCall.execute(OkHttpCall.java:190)
                                                                                        	at retrofit2.adapter.rxjava2.CallExecuteObservable.subscribeActual(CallExecuteObservable.java:45)
                                                                                        	at io.reactivex.Observable.subscribe(Observable.java:12267)
                                                                                        	at retrofit2.adapter.rxjava2.BodyObservable.subscribeActual(BodyObservable.java:34)
                                                                                        	at io.reactivex.Observable.subscribe(Observable.java:12267)
                                                                                        	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
                                                                                        	at io.reactivex.Observable.subscribe(Observable.java:12267)
                                                                                        	at io.reactivex.internal.operators.observable.ObservableFlatMap.subscribeActual(ObservableFlatMap.java:55)
                                                                                        	at io.reactivex.Observable.subscribe(Observable.java:12267)
                                                                                        	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
                                                                                        	at io.reactivex.Observable.subscribe(Observable.java:12267)
                                                                                        	at io.reactivex.internal.operators.observable.ObservableToListSingle.subscribeActual(ObservableToListSingle.java:58)
                                                                                        	at io.reactivex.Single.subscribe(Single.java:3603)
                                                                                        	at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
                                                                                        	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
                                                                                        	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
                                                                                        	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
                                                                                        	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
                                                                                        	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
                                                                                        	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
                                                                                        	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
                                                                                        	at java.lang.Thread.run(Thread.java:1012)

```

nicolas-raoul reviewed (changes requested):
Per https://github.com/commons-app/apps-android-commons/pull/6018#issuecomment-2534799789

angrezichatterbox commented:
I’ve implemented the suggested changes. For now, I’ve modified the description of the Toast Shows error feedback. I’ll make further changes based on your suggestions.

neeldoshii commented on app/src/main/java/fr/free/nrw/commons/navtab/MoreBottomSheetFragment.kt:
```diff
@@ -182,14 +182,17 @@ class MoreBottomSheetFragment : BottomSheetDialogFragment() {
-            }
+            }, { error ->
+                error.printStackTrace()
+                Toast.makeText(requireContext(), R.string.error_feedback, Toast.LENGTH_SHORT).show()
```
same

neeldoshii replied:
I guess one toast is left, can you change it to snackbar?

angrezichatterbox replied:
This was there earlier as well. Should I change it to Snackbar

neeldoshii replied:
Its okay for now.

neeldoshii commented on app/src/main/java/fr/free/nrw/commons/navtab/MoreBottomSheetFragment.kt:
```diff
@@ -182,14 +182,17 @@ class MoreBottomSheetFragment : BottomSheetDialogFragment() {
                 Toast.makeText(requireContext(), getString(messageResId), Toast.LENGTH_SHORT).show()
-            }
+            }, { error ->
+                error.printStackTrace()
```
Not safe to printStackTrace in direct logs, use Timber.e(). 

```diff
+ Timber.e(error)
- error.printStackTrace()
```

neeldoshii commented:
LGTM. Left few nitpicks in comments.

angrezichatterbox commented:
> LGTM. Left few nitpicks in comments.

Thanks, I have made the required changes :)

angrezichatterbox commented:
https://github.com/commons-app/apps-android-commons/blob/d7683f8ca6c41531ceeeec28501771c42bb7d00b/app/src/main/java/fr/free/nrw/commons/navtab/MoreBottomSheetFragment.kt#L192C1-L199C1

These lines were earlier using Toast as well. Should I change them to snackbar

nicolas-raoul commented:
> These lines were earlier using Toast as well. Should I change them to snackbar 

No need in this PR. This can be a separate PR. 🙂

nicolas-raoul reviewed (approved):
Great!
Unit tests passing on my machine, real tests went well.
Thanks!
